### PR TITLE
Strengthen NullableAnnotationExtensions.ToPublicAnnotation helper against null input.

### LINKED ISSUE #49754: NullReferenceException in Nullable Walker.
**Version Used**: 
   master: 808146464242d02f6c0ab596327eb49ef9ea13d8
   
**Steps to Reproduce**:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYAWAbgFgAofI4gOgBk0A7ARyupoGJWBXDBgjAMcAnFYixNGjlIEAKnFgAKEgGYAPKrYwC8WEgJ6CUAMYB7BHACUAPjNWbUWzQDeNAt6fWVDAGU4MXMYVSgCAF5HcIZDGGMoBgs/MwB3NBhzAAsvH09uah8CAF9bWy4SoA
```csharp
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

#nullable enable

namespace Repro
{
	public class Test
	{
		public void Test(IEnumerable<(int test, int score)> scores)
		{
			scores.Select(s => (s.test, s.score switch
			{

			}));
		}
	}
}
```

**Expected Behavior**:
No exception while parsing.

**Actual Behavior**:
Throw a NullReferenceException.
I tried to pinpoint the issue:  
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs#L636
Here `inferredType` is null, leading to  
`var inferredTypeWithAnnotations = TypeWithAnnotations.Create(inferredType);`  
having it's Type field being null too.  
At the end of this function `SetResult` call `SetAnalyzedNullability` 

It's last line do this:  
```csharp
_analyzedNullabilityMapOpt[expr] = (new NullabilityInfo(result.LValueType.ToPublicAnnotation(), result.RValueType.State.ToPublicFlowState()),
                                     // https://github.com/dotnet/roslyn/issues/35046 We're dropping the result if the type doesn't match up completely
                                     // with the existing type
    expr.Type?.Equals(result.RValueType.Type, TypeCompareKind.AllIgnoreOptions) == true ? result.RValueType.Type : expr.Type);
```
Which assume that the Type field is not null and throw there.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #49754.

### CODE REVIEW

**Review by RikkiGibson — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -142278,6 +142278,43 @@ static void F(object x, object? y)
...
+			scores.Select(s => (s.test, s.score switch
+			{
+
+			}));
+		}
+	}
+}"
+;
+            var comp = CreateCompilation(source0);
+            comp.VerifyEmitDiagnostics(
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142308]: >VerifyEmitDiagnostics [](start = 17, length = 21)

Do we call `ToPublicAnnotation()` here?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142308]: > Do we call ToPublicAnnotation() here?

Yes, while performing nullable analysis for the code

---
In reply to: [538670812](https://github.com/dotnet/roslyn/pull/49849#discussion_r538670812) [](ancestors = 538670812)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142308]: See this code in the walker:
```
                _analyzedNullabilityMapOpt[expr] = (new NullabilityInfo(result.LValueType.ToPublicAnnotation(), result.RValueType.State.ToPublicFlowState()),
                                                    // [https://github.com/dotnet/roslyn/issues/35046](https://github.com/dotnet/roslyn/issues/35046) We're dropping the result if the type doesn't match up completely
                                                    // with the existing type
                                                    expr.Type?.Equals(result.RValueType.Type, TypeCompareKind.AllIgnoreOptions) == true ? result.RValueType.Type : expr.Type);
```



---
In reply to: [538683287](https://github.com/dotnet/roslyn/pull/49849#discussion_r538683287) [](ancestors = 538683287,538670812)

**Comment by 333fred** [comment]
A more general question Aleksey: when I look at that bug, I wonder if the correct solution would be to have the switch expression have an error type, rather than null? It was unable to be inferred, so the behavior I would have expected would be to have an error there, not null.

**Comment by AlekseyTs** [comment]
> A more general question Aleksey: when I look at that bug, I wonder if the correct solution would be to have the switch expression have an error type, rather than null? It was unable to be inferred, so the behavior I would have expected would be to have an error there, not null.

In this particular scenario it looks like we are intentionally looking at unconverted switch expression node and null type for it is expected, I think, at least in some scenarios. We also have an explicit override to visit a node like that (an unconverted switch expression) with a comment indicating some of the scenarios when that might occur. For this specific repro, we explicitly get to the unconverted state, when we dig from converted tuple literal to its unconverted form. The converted tuple actually has converted switch expression that has error type as its type. In any case, I think making the change to the ToPublicAnnotation helper is the right thing to do regardless. If you feel like nullable walker has a problem specifically around handling switch expressions, I suggest you open a separate issue for that.

---
In reply to: [740825478](https://github.com/dotnet/roslyn/pull/49849#issuecomment-740825478) [](ancestors = 740825478)

**Pull request merged**